translator/source_code/cotrip_translator.py

```diff
@@ -0,0 +1,421 @@
+    description = alert.get('alert:Description')
+   
+    required_fields = [ polyline, coords, street, starttime, endtime, description]
+    for field in required_fields:
+        if not field:
+            logging.warning('Invalid alert with id =. Not all required fields are present')
+            return False
+            
+    try:
+        dateutil.parser.parse(alert['alert:ReportedTime'])
+        if alert.get('alert:ExpectedEndTime'):
+            dateutil.parser.parse(alert['alert:ExpectedEndTime'])
+    except ValueError:
+        logging.warning('Invalid alert with id . Invalid datetime format')
+        return False
+    
+    return True
+
+def reformat_datetime(datetime_string):
+    time = dateutil.parser.parse(datetime_string)
+    wzdx_format_datetime = time.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
+    return wzdx_format_datetime
+
+# Add ids to message
+#### This function may fail if some optional fields are not present (lanes, types_of_work, relationship, ...)
+def add_ids(message, add_ids):
+    if add_ids:
+        feed_info_id = message['road_event_feed_info']['feed_info_id']
+        data_source_id = message['road_event_feed_info']['data_sources'][0]['data_source_id']
+
+        road_event_length = len(message['features'])
+        road_event_ids = []
+        for i in range(road_event_length):
+            road_event_ids.append(str(uuid.uuid4()))
+
+        for i in range(road_event_length):
+            feature = message['features'][i]
+            road_event_id = road_event_ids[i]
+            feature['properties']['road_event_id'] = road_event_id
+            feature['properties']['data_source_id'] = data_source_id
+            feature['properties']['relationship']['relationship_id'] = str(uuid.uuid4())
+            feature['properties']['relationship']['road_event_id'] = road_event_id 
+
+    return message
+
+
+help_string = """ 
+
+Usage: python icone_translator.py [arguments]
+
+Global options:
+-h, --help                  Print this usage information.
+-i, --input                 specify the xml file to translate
+-o, --output                specify the output file for generated wzdx geojson message """
+
+def parse_arguments(argv):
+    inputfile = ''
+    outputfile = 'cotrip_wzdx_translated_output_message.geojson'
+    try:
+        opts, args = getopt.getopt(argv, "hi:o:", ["input=", "output="])
+    except getopt.GetoptError:
+        sys.exit(2)
+    for opt, arg in opts:
+        if opt in ("-h", "--help"):
+            print(help_string)
+            sys.exit()
+        elif opt in ("-i", "--input"):
+            inputfile = arg
+        elif opt in ("-o", "--output"):
+            outputfile = arg
+    return inputfile, outputfile
+inputfile, outputfile = parse_arguments(sys.argv[1:])
+
+
+def initialize_info():
+    info = {}
+
+    #### Consider whether this id needs to be hardcoded or generated
+    info['feed_info_id'] = "104d7746-688c-44ed-b195-2ee948bf9dfa"
```

Comment:
This should probably come from an environment setting, or config file. Currently this is the same id as in the icone_translator file, which means we are telling our applications the data is from the same location which is inaccurate.